sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriterV2.scala

```diff
@@ -343,3 +402,312 @@ trait CreateTableWriter[T] extends WriteConfigMethods[CreateTableWriter[T]] {
+   *     a subset of fields based on the provided assignment.
+   *   - `delete`: Delete all the target table records.
+   *
+   * @param condition a `Column` representing the condition to be evaluated for the action.
+   * @return a new `WhenNotMatchedBySource` object configured with the specified condition.
+   */
+  def whenNotMatchedBySource(condition: Column): WhenNotMatchedBySource[T]
+}
+
+/**
+ * A class for defining actions to be taken when matching rows in a DataFrame during
+ * an update operation.
+ *
+ * @param dfWriter   The DataFrameWriterV2 instance responsible for writing data to a
+ *                   target DataFrame.
+ * @param condition  An optional condition Expression that specifies when the actions
+ *                   should be applied.
+ *                   If the condition is None, the actions will be applied to all matched rows.
+ *
+ * @tparam T         The type of data in the DataFrame.
+ */
+case class WhenMatched[T] (dfWriter: DataFrameWriterV2[T], condition: Option[Expression]) {
+  /**
+   * Specifies an action to update all matched rows in the DataFrame.
+   *
+   * @return The DataFrameWriterV2 instance with the update all action configured.
+   */
+  def updateAll(): DataFrameWriterV2[T] = {
+    dfWriter.matchedActions = dfWriter.matchedActions :+ UpdateStarAction(condition)
+    this.dfWriter
+  }
+
+  /**
+   * Specifies an action to update matched rows in the DataFrame with the provided column
+   * assignments.
+   *
+   * @param set A Map of column names to Column expressions representing the updates to be applied.
+   * @return The DataFrameWriterV2 instance with the update action configured.
+   */
+  def update(set: Map[String, Column]): DataFrameWriterV2[T] = {
+    dfWriter.matchedActions = dfWriter.matchedActions :+
+      UpdateAction(condition, set.map(x => Assignment(expr(x._1).expr, x._2.expr)).toSeq)
+    this.dfWriter
+  }
+
+  /**
+   * Specifies an action to delete matched rows from the DataFrame.
+   *
+   * @return The DataFrameWriterV2 instance with the delete action configured.
+   */
+  def delete(): DataFrameWriterV2[T] = {
+    dfWriter.matchedActions = dfWriter.matchedActions :+ DeleteAction(condition)
+    this.dfWriter
+  }
+}
+
+/**
+ * A class for defining actions to be taken when no matching rows are found in a DataFrame
+ * during an update operation.
+ *
+ * @param dfWriter   The DataFrameWriterV2 instance responsible for writing data to a
+ *                   target DataFrame.
+ * @param condition  An optional condition Expression that specifies when the actions
+ *                   defined in this configuration should be applied.
+ *                   If the condition is None, the actions will be applied when there
+ *                   are no matching rows.
+ *
+ * @tparam T         The type of data in the DataFrame.
+ */
+case class WhenNotMatched[T] (dfWriter: DataFrameWriterV2[T], condition: Option[Expression]) {
```

Comment:
I checked the SQL grammar, WHEN NOT MATCHED only supports INSERT clause, which makes sense as there is no target row in the target table to update/delete.

Reply:
Update and Delete were removed